demo_haproxy/demo_haproxy.py

```diff
@@ -38,24 +64,51 @@ def generate_haproxy_config(primary_ports: List[int]) -> str:
     timeout http-request 10s
     log global
 
+# Tells HAProxy to also start the mirroring SPOA as a daemon when it starts up
+program mirror
+    command spoa-mirror --runtime 0 --address {SPOA_LISTEN_URL} --port {SPOA_LISTEN_PORT} --mirror-url {shadow_url}
+
+# This section surfaces a web form at localhost:8404 with statistics for the HAProxy instance
 frontend stats
     bind *:8404
     stats enable
     stats uri /
     stats refresh 10s
 
+# This section outlines the "frontend" for this instance of HAProxy and specifies the rules by which it receives
+# traffic
 frontend myfrontend
-    bind :80
+    bind :{HAPROXY_PORT}
+
+    # Set up the logging for the req/res stream to the primary cluster
     declare capture request len 80000
     declare capture response len 80000
     http-request capture req.body id 0
     log-format Request-URI:\\ %[capture.req.uri]\\nRequest-Method:\\ %[capture.req.method]\\nRequest-Body:\\ %[capture.req.hdr(0)]\\nResponse-Body:\\ %[capture.res.hdr(0)]
-    default_backend webservers
+    
+    # Associate this frontend with the primary cluster
+    default_backend primary_cluster
 
-backend webservers
+    # Required for the mirror SPOA to have access to the request
+    option http-buffer-request
+
+    # Tell HAProxy to also send traffic to this frontend to the mirror agent too
+    filter spoe engine mirror config /usr/local/etc/haproxy/mirror.conf
+
+# These are the primary Cluster's Nodes; traffic will be sent to them synchronously.  The default round robin LB
+# pattern is in effect.
+backend primary_cluster
     http-response capture res.body id 0
 {backend_server_section}
 
+# Mirror agents
+backend mirroragents
+    mode tcp
+    balance roundrobin
```

Comment:
Obviously this isn't somethign we're exploring right now, but does this line mean that if we had multiple shadow clusters it would round robin requests between them?

Reply:
If you stuck them all in the same `backend` block, then yes I'd imagine so.  However, you'd probably just make a  eparate `program` block that defines a separate mirror process, a separate `filter` line in the `frontend` block, and a separate `backend` block for it.  In other words - duplicate the full setup rather than just add a new `server` line to the existing `backend` block.  I think.  Maybe?